lldb/tools/lldb-dap/lldb-dap.cpp

```diff
@@ -3106,80 +3228,105 @@ void request_stackTrace(const llvm::json::Object &request) {
-                 i < (thread.GetNumFrames() +
-                      queue_backtrace_thread.GetNumFrames() + 1)) {
-        if (i == thread.GetNumFrames()) {
-          const uint32_t thread_idx =
-              queue_backtrace_thread.GetExtendedBacktraceOriginatingIndexID();
-          const char *queue_name = queue_backtrace_thread.GetQueueName();
-          auto name = llvm::formatv("Enqueued from {0} (Thread {1})",
-                                    queue_name, thread_idx);
-          stackFrames.emplace_back(
-              llvm::json::Object{{"id", thread.GetThreadID() + 1},
-                                 {"name", name},
-                                 {"presentationHint", "label"}});
-          continue;
-        }
-        frame = queue_backtrace_thread.GetFrameAtIndex(
-            i - thread.GetNumFrames() - 1);
-      } else if (exception_backtrace_thread.IsValid()) {
-        if (i == thread.GetNumFrames() +
-                     (queue_backtrace_thread.IsValid()
-                          ? queue_backtrace_thread.GetNumFrames() + 1
-                          : 0)) {
-          stackFrames.emplace_back(
-              llvm::json::Object{{"id", thread.GetThreadID() + 2},
-                                 {"name", "Original Exception Backtrace"},
-                                 {"presentationHint", "label"}});
-          continue;
-        }
+      // If we're out of extended backtraces, no more frames to load.
+      if (!frame.IsValid() && threadCluster.empty()) {
+        done = true;
+        break;
+      }
 
-        frame = exception_backtrace_thread.GetFrameAtIndex(
-            i - thread.GetNumFrames() -
-            (queue_backtrace_thread.IsValid()
-                 ? queue_backtrace_thread.GetNumFrames() + 1
-                 : 0));
+      // Between the thread and extended backtrace add a label.
+      if (offset != 0 && (i - offset) == 0) {
+        const uint32_t thread_idx =
+            current.GetExtendedBacktraceOriginatingIndexID();
+        const char *queue_name = current.GetQueueName();
+        std::string name;
+        if (queue_name != nullptr) {
+          name = llvm::formatv("Enqueued from {0} (Thread {1})", queue_name,
+                               thread_idx);
+        } else {
+          name = llvm::formatv("Thread {0}", thread_idx);
+        }
+        stackFrames.emplace_back(
+            llvm::json::Object{{"id", thread.GetThreadID() + 1},
+                               {"name", name},
+                               {"presentationHint", "label"}});
+      } else {
+        stackFrames.emplace_back(CreateStackFrame(frame));
       }
-      if (!frame.IsValid())
-        break;
-      stackFrames.emplace_back(CreateStackFrame(frame));
     }
 
-    body.try_emplace("totalFrames", totalFrames);
+    // If we loaded all the frames, set the total frame to the current total,
+    // otherwise use the totalFrames to indiciate more data is available.
+    body.try_emplace("totalFrames",
```

Comment:
Have you tested this? Since you are returning one more frame, I am not sure if this is causing VSCode client to ask for one more frame when you click "Get More frames" link. I believe @clayborg  suggested defining a page size (like 20 frames), then if it is not done, return  `startFrame + stackFrames.size() + STACK_PAGE_SIZE` instead. Then, client will ask for `STACK_PAGE_SIZE` instead of 1 more frame.

Reply:
Done.

I had tested this but the 20 page size does work better with deep stack traces.

Also updated the tests to verify this as well.